Fixed #20147 -- Added HttpRequest.headers.

santiagobasulto opened the pull request:
Original PR (with discussion included): https://github.com/django/django/pull/9925

I've rebased from master and squashed all the commits into one.

santiagobasulto commented:
👋 folks, there's only one build failing with: `ModuleNotFoundError: No module named 'memcache'`
I haven't changed any dependency or setup.py file, do you have an idea of what might be going on?

Thanks!

auvipy commented on docs/releases/2.1.txt:
```diff
@@ -211,6 +211,17 @@ Requests and Responses
   sets the ``Content-Disposition`` header to make the browser ask if the user
   wants to download the file. ``FileResponse`` also tries to set the
   ``Content-Type`` and ``Content-Length`` headers where appropriate.
+* Added :attr:`.HttpRequest.headers`.
```
shouldn't this moved to 2.2.txt?

auvipy replied:
you can drop this

santiagobasulto commented:
There seems to be an error with jenkins not finding the right commit after the rebase. Anybody has an idea how I could fix this?

auvipy commented:
everything seems passing

santiagobasulto commented:
You're right, when I submitted the PR originally, it was failing, it's working now 👍

santiagobasulto commented:
@carltongibson @timgraham PR finally rebased and working. Thanks for all your support. Gonna say this once again, this community is amazing.

jschneier commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,84 @@ def __getitem__(self, key):
+        yield tuple(elem)
+
+
+class ImmutableCaseInsensitiveDict(Mapping):
```
This isn't actually Immutable. I didn't trace far back enough to see if the WSGI `environ` is immutable but cursory inspection would indicate that `.META` is not actually immutable although it would probably be a bad idea to manipulate it (I might be wrong on this point).

carltongibson replied:
@santiagobasulto What are your thoughts on renaming this to `CaseInsensitiveMapping`? `Mapping` being the stdlib ABC you're implementing. (c.f. `ImmutableList.complain()`: we're taking measures there to avoid mutation. I don't think we want or need to get into any of that here.)

adamchainz commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,84 @@ def __getitem__(self, key):
+        return self
+
+
+class EnvironHeaders(ImmutableCaseInsensitiveDict):
```
this class has a bunch of details of WSGI environ so I think it more accurately belongs in `django.http.request`

santiagobasulto commented:
Hello @carltongibson @timgraham 👋 ! You think I could merge this one or should I make any other improvement?

carltongibson commented:
Hi @santiagobasulto. As far as I can see the comments from @jschneier and @adamchainz remain unaddressed. 

Just dropping `Immutable` from the naming may be enough to address @jschneier's point (???)

Overall it seems quite good to me. If you can address the points and then uncheck `Patch needs improvement` on the ticket, I'll have another look. (Seems like a nice feature for the amount of code.)

santiagobasulto commented:
Great, thanks @carltongibson. I'll get on it and update soon.

timgraham commented on docs/ref/request-response.txt:
```diff
@@ -135,7 +135,9 @@ All attributes should be considered read-only, unless stated otherwise.
 .. attribute:: HttpRequest.META
 
-    A dictionary containing all available HTTP headers. Available headers
+    A dictionary containing the `WSGI environ \
```
This change looks unrelated to the patch and should be submitted separately.

santiagobasulto replied:
I think that removing this and submitting it separately would lack the context later; what I did was just improving the docs a little bit more. From my experience, newcomers feel extremely confused when they find the prefix `HTTP_`.

carltongibson replied:
Just the PR is fine. 🙂

timgraham commented on docs/ref/request-response.txt:
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
     underscores in WSGI environment variables. It matches the behavior of
     Web servers like Nginx and Apache 2.4+.
 
+.. _headers:
```
This line isn't required.

timgraham commented on docs/ref/request-response.txt:
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
+    When accessing elements with ``headers.get``, ``headers[]`` or
+    checking membership (``in``), case is ignored. Examples::
+
+            'User-Agent' in request.headers  # True
```
This block is indent 4 spaces too much.

timgraham commented on tests/requests/tests.py:
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
                 self.assertEqual(request.build_absolute_uri(location=location), expected_url)
+
+
+class CaseInsensitiveHeadersTestCase(SimpleTestCase):
```
TestCase -> Tests

timgraham commented on docs/ref/request-response.txt:
```diff
@@ -135,7 +135,9 @@ All attributes should be considered read-only, unless stated otherwise.
 
-    A dictionary containing all available HTTP headers. Available headers
+    A dictionary containing the `WSGI environ \
+    <https://www.python.org/dev/peps/pep-0333/#environ-variables>`_, including all available HTTP headers.
```
Documentation should be wrapped at 79 characters.

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,247 @@ def f(x):
+        })
+
+    def test_create_dict_with_invalid_values(self):
+        with self.assertRaises(ValueError):
```
Use `assertRaisesMessage()`.

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,247 @@ def f(x):
+    def test_list(self):
+        self.assertEqual(
+            sorted(list(self.dict_1)),
+            sorted(['Accept', 'content-type']))
```
trailing ) goes on the next line (check your entire patch for this style)

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,247 @@ def f(x):
+        }))
+
+    def test_equals(self):
+        self.assertTrue(self.dict_1 == {
```
Why not assertEqual?

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,247 @@ def f(x):
+        self.assertEqual(self.dict_1['Content-type'], 'text/html')
+
+    def test_membership(self):
+        self.assertTrue('Accept' in self.dict_1)
```
assertIn

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -155,3 +155,247 @@ def f(x):
+            'HTTP_ACCEPT': '*',
+            'HTTP_HOST': 'example.com',
+            'HTTP_USER_AGENT': 'python-requests/1.2.0',
+
```
blank line not needed

timgraham commented on docs/ref/request-response.txt:
```diff
@@ -160,13 +162,38 @@ All attributes should be considered read-only, unless stated otherwise.
+            # .get():
+            request.headers.get('User-Agent')  # Returns the value of User-Agent
+            request.headers.get('user-agent')  # Returns the value of User-Agent
+
```
one blank line please

santiagobasulto replied:
There's only one line here Tim. You mean remove all the lines altogether?

carltongibson commented:
Hi @santiagobasulto. 

This still looks good. It would be a super addition for v2.2. Are you up for pushing it over the finish line? (If not we'll see if we can pick it up...) 

Re @jschneier's query on immutability, from [PEP 3333](https://www.python.org/dev/peps/pep-3333/#specification-details) (the WSGI spec):  

> The `environ` parameter is a dictionary object, containing CGI-style environment variables. This object **must** be a builtin Python dictionary (not a subclass, UserDict or other dictionary emulation), and the application is allowed to modify the dictionary in any way it desires.

So it's not immutable by design. 

Your `ImmutableCaseInsensitiveDict` implements `Mapping` (which is read-only, vs `MutableMapping`) so maybe just `CaseInsensitiveMapping`, not worrying about actual immutability. 
`Request.headers` is then (via your `EnvironHeaders`) a case-insensitive read-only accessor. (Just what we want.)

(Anybody wanting to actually mutate `Request.META` — Why? 🙂 — can access it directly.) 

Moving `EnvironHeaders` as per @adamchainz's review and making the adjustments Tim suggested shouldn't leaving it far off.

santiagobasulto commented:
Hey @carltongibson! Yes, I'll get to this immediately and push the fixes by tomorrow. Sorry, I've been quite busy lately but I don't want to keep postponing this.

zachborboa commented on tests/requests/tests.py:
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
+        self.assertEqual(dict(request.headers), {
+            'Content-Type': 'text/html',
+            'Content-Length': '100',
+            'Accept': '*', 'Host': 'example.com',
```
```suggestion
            'Accept': '*',
            'Host': 'example.com',
```

santiagobasulto replied:
Thanks @zachborboa!

zachborboa commented on tests/requests/tests.py:
```diff
@@ -806,3 +806,121 @@ def test_request_path_begins_with_two_slashes(self):
+        self.assertEqual(dict(request.headers), {
+            'Content-Type': 'text/html',
+            'Content-Length': '100',
+            'Accept': '*', 'Host': 'example.com',
```
```suggestion
            'Accept': '*',
            'Host': 'example.com',
```

carltongibson commented:
Hi Santiago. Thanks for the work! I will review on Tuesday, so no panic. 🙂

santiagobasulto commented:
@carltongibson Agree completely on the Mapping suggestion, I'll fix it now. I somehow missed one of Tim fixes. I'll update everything together.

maguayo commented:
Looking good! I love the changes

santiagobasulto commented:
@carltongibson all done now, I'll submit the doc changes in a separate PR. I'll take some time to read the whole text and see if there are other improvements to do. Thank so much again for your patience.

santiagobasulto commented:
@carltongibson @timgraham hey folks 👋! Is there anything else you'd like me to change here? I 've addressed all your comments and seems like everything is passing. January 14th is the deadline for 2.2 features, right?

timgraham commented:
Don't forget to uncheck "Patch needs improvement" on the ticket so that it appears in the review queue.

santiagobasulto commented:
@timgraham I forgot about that 🤦‍♂️, thanks for the help.

collinanderson commented:
@santiagobasulto If you get a chance, do you want to update the doc string for CaseInsensitiveMapping with the text that carltongibson suggested?

> Mapping allowing case-insensitive key lookups. Original case of keys is preserved for iteration and string representation.

Otherwise, I like this PR. Thanks so much! In same ways it would be nice if you changed one dict it would change the other, but I don't think that's going to be an issue.

collinanderson commented:
@santiagobasulto yeah I don't think updating headers is worth it.

timgraham commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,63 @@ def __getitem__(self, key):
+def _destruct_iterable_mapping_values(data):
+    for i, elem in enumerate(data):
+        if len(elem) != 2:
+            raise ValueError("dictionary update sequence element #{} has "
```
Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

timgraham commented on django/http/request.py:
```diff
@@ -572,3 +578,38 @@ def validate_host(host, allowed_hosts):
+
+
+class EnvironHeaders(CaseInsensitiveMapping):
+    IGNORE_EXCEPTIONS = {'HTTP_CONTENT_TYPE', 'HTTP_CONTENT_LENGTH'}
```
I think this deserves a comment. Why are these headers "ignored" (whatever that means). Why are the headers below "special unchanged" (whatever that means).

collinanderson replied:
Maybe just reference that these don't get the prefix as specified in https://www.python.org/dev/peps/pep-0333/

I understand why `SPECIAL_UNCHANGED_HEADERS` is needed but I don't understand why `IGNORE_EXCEPTIONS` is needed. Do we ever get incoming `HTTP_CONTENT_TYPE` keys?

santiagobasulto replied:
I remember I did some research about the possible variations of multiple WSGI servers and some users noticed that some servers were not honoring PEP-333 and included `HTTP_CONTENT_TYPE`. 

I found that Python's [`wsgiref.validate`](https://docs.python.org/3.6/library/wsgiref.html) module [raises an exception](https://github.com/python/cpython/blob/master/Lib/wsgiref/validate.py#L310) if ever found.

Options:

a) leave it as it is: A robust alternative. If someone is using a poorly designed wsgi server, it'll still work. Maybe issue a warning.
b) Remove it all together. Assume that wsgi implementations will honor PEP 333.
c) Make it fail if ever found one (raise an exception). Similar to what `wsgiref.validate` is doing. This is kind of redundant, i don't like it TBH.

collinanderson replied:
Ok, yeah it looks like uwsgi might have, at least at some point, both CONTENT_TYPE and HTTP_CONTENT_TYPE.  So, yeah, I think ignoring HTTP_CONTENT_TYPE and trusting only CONTENT_TYPE is a good idea (a).

santiagobasulto replied:
@timgraham all points of your review addressed and I added the comments on these variables, just in case.

timgraham replied:
I don't think Django has existing workarounds for this issue, e.g. https://github.com/django/django/blob/f021c110d02fd7ca32ae56f511b46e5d138b6c73/django/core/handlers/wsgi.py#L83
I think it adds necessary complexity.

santiagobasulto replied:
@timgraham 👍

timgraham commented on django/utils/datastructures.py:
```diff
@@ -280,3 +281,63 @@ def __getitem__(self, key):
+
+    def __eq__(self, other):
+        if not isinstance(other, Mapping):
+            return NotImplemented
```
I'd expect to see a test for this. Please audit test coverage carefully and make sure all lines are covered.

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -148,3 +148,121 @@ def f(x):
+        })
+
+    def test_repr(self):
+        dict_1 = CaseInsensitiveMapping({
```
Use a single line throughout where possible.

timgraham commented on tests/utils_tests/test_datastructures.py:
```diff
@@ -148,3 +148,121 @@ def f(x):
+    def test_delitem(self):
+        """del should raise a TypeError because this is immutable"""
+        # Preconditions
+        self.assertTrue('Accept' in self.dict_1)
```
assertIn

santiagobasulto commented on docs/ref/request-response.txt:
```diff
@@ -160,13 +160,37 @@ All attributes should be considered read-only, unless stated otherwise.
     underscores and adding an ``HTTP_`` prefix to the name. So, for example, a
     header called ``X-Bender`` would be mapped to the ``META`` key
     ``HTTP_X_BENDER``.
+    See :attr:`~HttpRequest.headers` for more convenient header
```
Does "more intuitive" sound nicer? There's an example after it `request.headers['User-Agent']`, which makes it more obvious, or adds extra context.

santiagobasulto commented:
Hey folks 👋 , just finished addressing comments by @timgraham; also improved docs with the suggestions by @carltongibson (replace "dict" with case insensitive mapping).
The only open thing to resolve is in this comment: https://github.com/django/django/pull/10171#discussion_r246975198

I have the whole day free tomorrow (traveling to Raleigh NC for the first time, if you have recommendations of cool things to do they're welcome 😁). So I'd like to finish every open issue on this one and get it merged between the feature freeze. IMO having `headers` will really help beginners.

chicheng commented:
@santiagobasulto Could you document what will be returned if duplicated header present in the request?

https://tools.ietf.org/html/rfc7230#section-3.2.2
>    A sender MUST NOT generate multiple header fields with the same field name in a message unless either the entire field value for that header field is defined as a comma-separated list [i.e., #(values)] or the header field is a well-known exception (as noted below).

timgraham commented on django/http/request.py:
```diff
@@ -572,3 +578,43 @@ def validate_host(host, allowed_hosts):
+
+    @classmethod
+    def _style_header_name(cls, header_name):
+        return header_name.title()
```
Does this need to be a separate method? Seems unnecessary to me.

santiagobasulto replied:
This can be extended by subclasses and overwritten, that's why I defined it as a different method.

timgraham commented on tests/requests/tests.py:
```diff
@@ -830,3 +830,251 @@ def test_request_path_begins_with_two_slashes(self):
+    def test_parse_cgi_headers_basics(self):
+        parser = EnvironHeaders.parse_cgi_header_name
+
+        self.assertEqual(parser('HTTP_ACCEPT'), 'Accept')
```
Similar to my "use a simpler dictionary" comment above, this fells like overkill. The logic is fairly simple. What does "'HTTP_X_CUSTOM_HEADER_1" test that "HTTP_X_CUSTOM_HEADER_2" doesn't? Also you can use a loop with `subTest()` to simplify rather than have separate methods. For example:
https://github.com/django/django/blob/f021c110d02fd7ca32ae56f511b46e5d138b6c73/tests/utils_tests/test_http.py#L156-L158

adamchainz commented:
🎉
